### PR TITLE
WT-5796 Speed up verification of file's associated history store information.

### PR DESCRIPTION
**Pull request by keithbostic**
@luke-pearson, @sulabhM, Jeremy Tay did the original work and you were the reviewers; of course, please turf the review elsewhere if that's appropriate.

I haven't tested if this change fixes the timeouts in Evergreen tests, but the speedup is more than 100x, and is roughly the speed of verify in previous releases.

The 903469b commit is the interesting one, and 1c4b3cc adds some minor additional changes (the other commits are code shuffling).

### CODE REVIEW

**Comment by agorrod** [comment]
@luke-pearson and @sulabhM could you please take a look at this PR.

File: src/history/hs.c [hunk 1]
```diff
@@ -1239,3 +1239,123 @@ __hs_delete_key_from_pos(
...
+                continue;
+        }
+        WT_ERR(__wt_buf_set(session, prev_hs_key, hs_key->data, hs_key->size));
+        prev_btree_id = btree_id;
+
+        /* Re-purpose the previous-key buffer on error, safe because we're about to error out. */
+        data_cursor->set_key(data_cursor, hs_key);
+        if ((ret = data_cursor->search(data_cursor)) == WT_NOTFOUND)
+            WT_ERR_MSG(session, ret,
+              "In %s, the associated history store key %s not found in the data store", uri_itr,
```
Comment (luke-pearson) [inline src/history/hs.c:1347]: Maybe "was not found" instead of just not found.

File: src/btree/bt_vrfy.c [hunk 2]
```diff
@@ -1347,6 +1066,36 @@ __verify_page_cell(
...
+                WT_RET(__wt_debug_key_value(session, vs->tmp1, WT_RECNO_OOB, 0, &unpack));
+#endif
+        } else if (page->type == WT_PAGE_COL_VAR) {
+            rle = __wt_cell_rle(&unpack);
+            p = vs->tmp1->mem;
+            WT_RET(__wt_vpack_uint(&p, 0, recno));
+            vs->tmp1->size = WT_PTRDIFF(p, vs->tmp1->mem);
+            WT_RET(__verify_key_hs(session, vs->tmp1, unpack.start_ts, vs));
+
+#ifdef HAVE_DIAGNOSTIC
```
Comment (luke-pearson) [inline src/btree/bt_vrfy.c:1092]: Is it possible to maintain consistency and have the __wt_debug_key_value call prior to the __verify_key_hs call?

File: src/btree/bt_vrfy.c [hunk 3]
```diff
@@ -13,11 +13,12 @@
  * prettier.
  */
 typedef struct {
-    uint64_t record_total; /* Total record count */
+    uint64_t records_so_far; /* Total record count seen so far */
```
Comment (luke-pearson) [inline src/btree/bt_vrfy.c:16]: Record count seen so far, seems like a better comment as total and count are effectively the same thing here.

File: src/btree/bt_vrfy.c [hunk 4]
```diff
@@ -1168,54 +826,115 @@ __verify_txn_addr_cmp(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t cell_num,
...
+     * in which case we initialize our newest with the max txn-id.
+     */
+    older_stop_ts = 0;
+
+    /*
+     * Open a history store cursor positioned at the end of the data store key (the newest record)
+     * and iterate backwards until we reach a different key or btree.
+     */
+    hs_cursor = session->hs_cursor;
+    hs_cursor->set_key(hs_cursor, hs_btree_id, tmp1, WT_TS_MAX, UINT64_MAX);
```
Comment (luke-pearson) [inline src/btree/bt_vrfy.c:861]: Does UINT64_MAX here need to be WT_TXN_MAX?

File: src/history/hs.c (hunk 1 above)
Comment (keithbostic) [reply src/history/hs.c]: Pushed in 66ccd6594.

File: src/btree/bt_vrfy.c (hunk 3 above)
Comment (keithbostic) [reply src/btree/bt_vrfy.c]: Pushed in 66ccd6594.

File: src/btree/bt_vrfy.c (hunk 4 above)
Comment (keithbostic) [reply src/btree/bt_vrfy.c]: Good catch, that was in the original code and I didn't really look at it -- I think you're correct, we should use `WT_TXN_MAX`.

Pushed in 66ccd6594.

File: src/btree/bt_debug.c [hunk 5]
```diff
@@ -803,17 +804,16 @@ __wt_debug_key_value(WT_SESSION_IMPL *session, WT_ITEM *key, WT_CELL_UNPACK *val
...
+    else
         WT_ERR(__debug_item_key(ds, "K", key->data, key->size));
-    if (value != NULL) {
-        WT_ERR(__debug_time_pairs(
-          ds, "T", value->start_ts, value->start_txn, value->stop_ts, value->stop_txn));
-        WT_ERR(__debug_cell_data(ds, NULL, value != NULL ? value->type : 0, "V", value));
-    }
+    WT_ERR(__debug_time_pairs(
+      ds, "T", value->start_ts, value->start_txn, value->stop_ts, value->stop_txn));
+    WT_ERR(__debug_cell_data(ds, NULL, value != NULL ? value->type : 0, "V", value));
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_debug.c:813]: Yeah, that was deliberate.

For some reason, this function was added with a check of value against NULL, but none of its callers ever pass a NULL value.

[the remark this answers is not in the document]
Comment (sulabhM) [reply src/btree/bt_debug.c:813]: Cool, good to go then.

**Pull request merged**